core/src/ast/methods.rs

```diff
@@ -110,44 +134,22 @@ impl Method {
     pub fn borrowed_params(&self) -> BorrowedParams {
         // To determine which params the return type is bound to, we just have to
         // find the params that contain a lifetime that's also in the return type.
-        if let Some(ref return_type) = self.return_type {
-            // The lifetimes that must outlive the return type
-            let lifetimes = {
-                let mut return_type_lifetimes: Vec<&NamedLifetime> = vec![];
-
-                return_type.visit_lifetimes(&mut |lifetime, _| -> ControlFlow<()> {
-                    match lifetime {
-                        Lifetime::Named(named) => return_type_lifetimes.push(named),
-                        Lifetime::Anonymous => {
-                            // TODO(160): Once lifetime elision in return types
-                            // is allowed, we can remove this.
-                            // This is also caught in the validity check, so this should
-                            // never happen.
-                            panic!("Anonymous lifetimes not yet allowed in return types")
+        if let Some(borrowed_lifetimes) = self.borrowed_lifetimes() {
+            let borrows_self = self
+                .self_param
+                .as_ref()
+                .map(|self_param| {
+                    // Check if `self` is a reference with a lifetime in the return type.
+                    if let Some((Lifetime::Named(ref name), _)) = self_param.reference {
```

Comment:
I'm not sure if you mean the case where the `Self` type could expand to `Foo<'a, 'b>`, because that case is handled 5 lines down
```rust
self_param.path_type.lifetimes.iter().any(|lt| {
    matches!(lt, Lifetime::Named(name) if borrowed_lifetimes.contains(&name))
})
```